### ISSUE TITLE
Enforce Optimal Variable Declaration Scope

### ISSUE THREAD

**Issue by tatarysh** [eslint/eslint#19301]
### Rule details

Analyze variable declarations within blocks, ensuring they are not declared prematurely.

### Related ECMAScript feature

This rule relates to modern scoping introduced with let and const in ES6. These block-scoped declarations are designed to improve variable management, and this rule builds upon their usage to encourage best practices for scope management.

### What type of rule is this?

Suggests an alternate way of doing something

### Example code

```js
Invalid

if (isConfigurable(currentProduct)) {
  const prices = (currentProduct.children ?? []).map((child) => child.real_price);

  const minPrice = Math.min(...prices));

  if (prices.length) {
    return prices.find((price) => price.amount === minPrice);
  }
}


Valid:

if (isConfigurable(currentProduct)) {
  const prices = (currentProduct.children ?? []).map((child) => child.real_price);

  if (prices.length) {
    const minPrice = Math.min(...prices));
    return prices.find((price) => price.amount === minPrice);
  }
}
```

### Why should this rule be in the core instead of a plugin?

This proposal introduces a new ESLint rule to enforce declaring variables in the most appropriate and minimal scope, ensuring they are only accessible within the logical context in which they are required. The rule aims to improve code clarity, prevent premature declarations, and reduce potential bugs caused by unintended variable access or misuse.

### Participation

- [ ] I am willing to submit a pull request to implement this rule.

**Comment by nzakas** [comment]
Thanks for the suggestion. If I'm reading this correctly, you're looking for a rule that would ensure the smallest possible scope for any given variable?

A couple of concerns:

1. This type of analysis can be a bit expensive, as we'd need to search through all references for each variable.
2. I'm not sure we could safely autofix such a problem as it could require moving around a significant amount of code.

I'm not opposed to this rule, but there is a cost-benefit question in my mind.

@eslint/eslint-team thoughts?

**Comment by tatarysh** [comment]
yes, you understand me correctly.

I think that would be a useful rule - but this is what I thought, it can be quite complicated and not cost-effective, although I allowed myself to create this issue to know your opinion.

**Comment by mdjermanovic** [comment]
I'm not sure if this rule can be reliably implemented because moving the variable declaration also means moving the initializer to a different point of execution, which might not be possible because of the data the initializer uses or side effects it produces, so the rule might end up with a lot of false positives.

Example 1:

```js
/** This function gets the first two numbers from a sequence and returns their sum if the second one is positive */
function sum(numbers) {
    const first = numbers.get();
    const second = numbers.get();
    if (second > 0) {
        // can't move `const first = numbers.get();` here
        return first + second;
    }
}
```

Example 2:

```js
/** This function always runs `doSomething()` and optionally checks the outcome */
function run(check) {
    const result = doSomething(); // do some side effects
    if (check) {
        // can't move `const result = doSomething();` here
        checkResult(result);
    }
}
```

**Comment by JoshuaKGoldberg** [comment]
Agreed, side effects make this much less reliable as a rule. I think a plugin like [`eslint-plugin-functional`](https://github.com/eslint-functional/eslint-plugin-functional) would be much more capable of enforcing such a practice.

I'm -1 on this rule in ESLint core but +1 on it existing in a plugin that relies on there not being side effects. 🙂

**Comment by nzakas** [comment]
It looks like there's consensus that this isn't appropriate for the core, so closing.
